src/NLog.Extensions.Logging/Logging/NLogLogger.cs

```diff
@@ -36,52 +37,224 @@ public void Log<TState>(Microsoft.Extensions.Logging.LogLevel logLevel, EventId
+                {
+                    return false;
+                }
             }
-            return LogEventInfo.Create(nLogLogLevel, _logger.Name, message);
+
+            return true;
         }
 
-        private static bool IsNonDigitValue(string value)
+        private static object[] CreateStructuredLogEventInfoParameters(NLogMessageParameterList messageParameters, MessageTemplateParameters messagetTemplateParameters, out List<MessageTemplateParameter> extraProperties)
         {
-            return !String.IsNullOrEmpty(value) && (value.Length != 1 || !Char.IsDigit(value[0]));
+            extraProperties = null;
+
+            var paramsArray = new object[messagetTemplateParameters.Count];
+            int startPos = 0;
+            for (int i = 0; i < messageParameters.Count; ++i)
+            {
+                bool extraProperty = true;
+                for (int j = startPos; j < messagetTemplateParameters.Count; ++j)
+                {
+                    if (messageParameters[i].Name == messagetTemplateParameters[j].Name)
+                    {
+                        extraProperty = false;
+                        paramsArray[j] = messageParameters[i].Value;
+                        if (startPos == j)
+                            startPos++;
+                        break;
+                    }
+                }
+
+                if (extraProperty)
+                {
+                    extraProperties = extraProperties ?? new List<MessageTemplateParameter>();
+                    extraProperties.Add(messageParameters[i]);
+                }
+            }
+
+            return paramsArray;
+        }
+
+        private static object[] CreatePositionalLogEventInfoParameters(NLogMessageParameterList messageParameters, MessageTemplateParameters messagetTemplateParameters, out List<MessageTemplateParameter> extraProperties)
+        {
+            extraProperties = null;
+
+            var maxIndex = FindMaxIndex(messagetTemplateParameters);
+            object[] paramsArray = null;
+            for (int i = 0; i < messageParameters.Count; ++i)
+            {
+                // First positional name is the startPos
+                if (char.IsDigit(messageParameters[i].Name[0]) && paramsArray == null)
+                {
+                    paramsArray = new object[maxIndex + 1];
+                    for (int j = 0; j <= maxIndex; ++j)
+                    {
+                        if (i + j < messageParameters.Count)
+                            paramsArray[j] = messageParameters[i + j].Value;
+                    }
+                    i += maxIndex;
+                }
+                else
+                {
+                    extraProperties = extraProperties ?? new List<MessageTemplateParameter>();
+                    extraProperties.Add(messageParameters[i]);
+                }
+            }
+
+            return paramsArray ?? new object[maxIndex + 1];
```

Comment:
we could do `new object[maxIndex + 1]` at the declaration? (line 217) (and remove line 223)

Reply:
Well I (ab)use it for only having one location for adding extra properties. But I guess it could be replaced with with an extra local bool-parameter.